File: src/p141/linked_list_cycle.py
from __future__ import annotations
from typing import Optional

class ListNode:
    def __init__(self, x: int, node: Optional[ListNode]) -> None:
        self.val = x
        self.next = node


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # edge cases for fast pointer
        if not head or not head.next or not head.next.next:
            return False
        
        fast = head
        slow = head

        while slow.next and fast and fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                return True

        return False

# this one is awkward to test

File: src/p141/linked_list_cycle.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
   public:
    bool hasCycle(ListNode *head) {
        if (!head) {
            return false;
        }
        ListNode *slow = head, *fast = head->next;
        while (fast && fast->next && fast->next->next) {
            if (slow == fast) {
                return true;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
        return false;
    }
};
